File: ios/VMflow/Models/Deal.swift
import Foundation

// MARK: - API Response

struct DealSearchResponse: Codable {
    let deals: [Deal]
    let fromCache: Bool
    let searchedProducts: Int?
    let totalDeals: Int?
}

// MARK: - Deal Model

struct Deal: Codable, Identifiable {
    let id: UUID
    let productId: UUID?
    let keywordId: UUID?
    let matchedTerm: String?
    let retailer: String
    let dealTitle: String
    let dealPrice: Double?
    let regularPrice: Double?
    let discountPct: Double?
    let validFrom: String?
    let validUntil: String?
    let imageUrl: String?
    let imageUrlLarge: String?
    let sourceUrl: String?
    let externalUrl: String?
    let matchedBy: String
    let confidence: Double
    let matchedTokens: [String]?
    let requiresApp: Bool
    let fetchedAt: String
    let offerId: String?
    let products: DealProduct?
    let dealKeywords: DealKeywordMatch?

    enum CodingKeys: String, CodingKey {
        case id, retailer, confidence, products
        case productId = "product_id"
        case keywordId = "keyword_id"
        case matchedTerm = "matched_term"
        case dealTitle = "deal_title"
        case dealPrice = "deal_price"
        case regularPrice = "regular_price"
        case discountPct = "discount_pct"
        case validFrom = "valid_from"
        case validUntil = "valid_until"
        case imageUrl = "image_url"
        case imageUrlLarge = "image_url_large"
        case sourceUrl = "source_url"
        case externalUrl = "external_url"
        case matchedBy = "matched_by"
        case matchedTokens = "matched_tokens"
        case requiresApp = "requires_app"
        case fetchedAt = "fetched_at"
        case offerId = "offer_id"
        case dealKeywords = "deal_keywords"
    }

    // MARK: - Formatted Properties

    var formattedDealPrice: String? {
        guard let price = dealPrice else { return nil }
        return String(format: "%.2f \u{20AC}", price)
    }

    var formattedRegularPrice: String? {
        guard let price = regularPrice else { return nil }
        return String(format: "%.2f \u{20AC}", price)
    }

    var formattedDiscount: String? {
        guard let pct = discountPct, pct > 0 else { return nil }
        return "-\(Int(pct))%"
    }

    var productName: String {
        if let name = products?.name { return name }
        if let label = dealKeywords?.label, !label.isEmpty { return label }
        if let firstTerm = dealKeywords?.terms?.first, !firstTerm.isEmpty { return firstTerm }
        if let firstLinked = dealKeywords?.linkedProducts.first?.name { return firstLinked }
        if let term = matchedTerm, !term.isEmpty { return term }
        return "Unknown Product"
    }

    var productImagePath: String? {
        products?.imagePath ?? dealKeywords?.linkedProducts.first?.imagePath
    }

    var productSellprice: Double? {
        products?.sellprice ?? dealKeywords?.linkedProducts.first?.sellprice
    }

    // MARK: - Validity

    enum ValidityStatus {
        case upcoming, active, expiring, expired

        var label: String {
            switch self {
            case .upcoming: return String(localized: "Not valid yet")
            case .active: return String(localized: "Active")
            case .expiring: return String(localized: "Expiring soon")
            case .expired: return String(localized: "Expired")
            }
        }
    }

    /// Calendar-day validity, mirroring the web's `app/lib/dealValidity.ts` —
    /// keep both in sync. `valid_from`/`valid_until` are DATE columns holding
    /// calendar days in the company's timezone, so they're parsed as local
    /// days. The upcoming check deliberately doesn't require an end date: a
    /// deal without `valid_until` used to read as `.active` even when it only
    /// started next week.
    var validityStatus: ValidityStatus {
        let today = Calendar.current.startOfDay(for: Date())
        if let until = validUntilDate, until < today { return .expired }
        if let from = validFromDate, from > today { return .upcoming }
        if let until = validUntilDate, Self.days(from: today, to: until) <= 2 { return .expiring }
        return .active
    }

    var isValid: Bool {
        guard let until = validUntilDate else { return true }
        return until >= Calendar.current.startOfDay(for: Date())
    }

    /// True when the deal can be bought today (not upcoming, not expired).
    var isValidToday: Bool {
        validityStatus == .active || validityStatus == .expiring
    }

    var validFromDate: Date? { validFrom.flatMap(Self.parseDay) }
    var validUntilDate: Date? { validUntil.flatMap(Self.parseDay) }

    /// Whole days until the deal starts (≥ 1), only while it is upcoming.
    var startsInDays: Int? {
        guard validityStatus == .upcoming, let from = validFromDate else { return nil }
        return Self.days(from: Calendar.current.startOfDay(for: Date()), to: from)
    }

    var formattedValidUntil: String? {
        validUntilDate.map { Self.displayFormatter.string(from: $0) }
    }

    var formattedValidFrom: String? {
        validFromDate.map { Self.displayFormatter.string(from: $0) }
    }

    /// "valid from Mon, 28.09." — weekday first, since "is it Monday yet?"
    /// is the question when deciding whether to drive to the store.
    var validFromLabel: String? {
        validFromDate.map { String(format: String(localized: "valid from %@"), Self.shortDay($0)) }
    }

    /// "tomorrow" / "in 3 days".
    var startsInLabel: String? {
        startsInDays.map(Self.relativeDaysLabel)
    }

    static func relativeDaysLabel(_ days: Int) -> String {
        days == 1
            ? String(localized: "tomorrow")
            : String(format: String(localized: "in %lld days"), days)
    }

    static func shortDay(_ date: Date) -> String {
        date.formatted(.dateTime.weekday(.abbreviated).day(.twoDigits).month(.twoDigits))
    }

    static func longDay(_ date: Date) -> String {
        date.formatted(.dateTime.weekday(.wide).day().month(.wide))
    }

    /// Sort rank for "valid today first, then upcoming, then expired".
    var validityRank: Int {
        switch validityStatus {
        case .active, .expiring: return 0
        case .upcoming: return 1
        case .expired: return 2
        }
    }

    // MARK: - Confidence

    enum ConfidenceLevel {
        case high, medium, low

        init(value: Double) {
            if value >= 0.85 { self = .high }
            else if value >= 0.65 { self = .medium }
            else { self = .low }
        }
    }

    var confidenceLevel: ConfidenceLevel {
        ConfidenceLevel(value: confidence)
    }

    var formattedConfidence: String {
        "\(Int(confidence * 100))%"
    }

    // MARK: - Date Helpers

    private static let dateFormatter: DateFormatter = {
        let f = DateFormatter()
        f.calendar = Calendar(identifier: .gregorian)
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = .current
        f.dateFormat = "yyyy-MM-dd"
        return f
    }()

    private static let displayFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateStyle = .medium
        f.timeStyle = .none
        return f
    }()

    /// Local midnight of a `YYYY-MM-DD` value (tolerates a timestamp suffix).
    private static func parseDay(_ value: String) -> Date? {
        dateFormatter.date(from: String(value.prefix(10)))
    }

    private static func days(from: Date, to: Date) -> Int {
        Calendar.current.dateComponents([.day], from: from, to: to).day ?? 0
    }
}

// MARK: - Deal Product (joined relation)

struct DealProduct: Codable {
    let id: UUID?
    let name: String?
    let imagePath: String?
    let sellprice: Double?

    enum CodingKeys: String, CodingKey {
        case id, name, sellprice
        case imagePath = "image_path"
    }
}

// MARK: - Deal Keyword (joined relation for keyword-match rows)

/// Keyword group info embedded via PostgREST nested select when a deal was
/// matched against a user-defined keyword group rather than a single product.
/// The `deal_keyword_products` side returns rows like `{ products: {...} }`,
/// which we flatten into `linkedProducts` for convenient access.
struct DealKeywordMatch: Codable {
    let id: UUID?
    let label: String?
    let terms: [String]?
    let linkedProducts: [DealProduct]

    private struct LinkedProductRow: Codable {
        let products: DealProduct?
    }

    enum CodingKeys: String, CodingKey {
        case id, label, terms
        case dealKeywordProducts = "deal_keyword_products"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decodeIfPresent(UUID.self, forKey: .id)
        label = try c.decodeIfPresent(String.self, forKey: .label)
        terms = try c.decodeIfPresent([String].self, forKey: .terms)
        let rows = try c.decodeIfPresent([LinkedProductRow].self, forKey: .dealKeywordProducts) ?? []
        linkedProducts = rows.compactMap { $0.products }
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encodeIfPresent(id, forKey: .id)
        try c.encodeIfPresent(label, forKey: .label)
        try c.encodeIfPresent(terms, forKey: .terms)
        let rows = linkedProducts.map { LinkedProductRow(products: $0) }
        try c.encode(rows, forKey: .dealKeywordProducts)
    }
}

// MARK: - Deal User State (archived / pinned per user)

/// Per-user annotation on a deal, keyed by (retailer, offer_id) on the server.
/// Rows only exist for deals the user has interacted with; absence of a row
/// means not-archived and not-pinned.
struct DealUserStateRow: Codable {
    let retailer: String
    let offerId: String
    let archivedAt: String?
    let pinnedAt: String?

    enum CodingKeys: String, CodingKey {
        case retailer
        case offerId = "offer_id"
        case archivedAt = "archived_at"
        case pinnedAt = "pinned_at"
    }
}

// MARK: - Deduplicated Deal

/// Collapses multiple deal_cache rows that share the same (retailer, offer_id)
/// into a single card. Brand-wide offers that fuzzy-matched many catalog
/// entries (e.g. "Red Bull versch. Sorten" matching every Red Bull variant)
/// used to render as N near-identical duplicates; now they appear as one card
/// with the full list of matched products accessible in the detail sheet.
struct DedupedDeal: Identifiable {
    /// Stable cross-refresh key — matches the server-side `deal_user_state`
    /// composite key (retailer, offer_id). Safe to use as SwiftUI Identifiable.
    let key: String
    let retailer: String
    let offerId: String
    let primary: Deal
    let matchedProducts: [MatchedProduct]
    let matchedKeywords: [DealKeywordMatch]
    let archived: Bool
    let pinned: Bool
    let pinnedAt: String?

    var id: String { key }

    struct MatchedProduct: Identifiable {
        let id: UUID
        let name: String
        let imagePath: String?
        let sellprice: Double?
        let confidence: Double
    }

    /// Forwards through to the underlying primary deal so existing UI helpers
    /// (formattedDealPrice, validityStatus, confidenceLevel, …) keep working
    /// without needing a parallel set of DedupedDeal computed properties.
    var dealTitle: String { primary.dealTitle }
    var requiresApp: Bool { primary.requiresApp }
    var discountPct: Double? { primary.discountPct }
    var confidence: Double { primary.confidence }
    var imageUrl: String? { primary.imageUrl }
    var imageUrlLarge: String? { primary.imageUrlLarge }
}

File: android/app/src/main/java/xyz/vmflow/models/Deal.kt
package xyz.vmflow.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Response of the `deal-search` edge function. Mirrors iOS `DealSearchResponse`
 * (`ios/VMflow/Models/Deal.swift`). `message` carries "No deal-source providers
 * enabled" when the company has no active provider (same check as the web's
 * `useDeals.noProviders`).
 */
@Serializable
data class DealSearchResponse(
    val deals: List<Deal> = emptyList(),
    val fromCache: Boolean = false,
    val searchedProducts: Int? = null,
    val totalDeals: Int? = null,
    val message: String? = null,
)

/** One `deal_cache` row as returned by `deal-search`. Mirrors iOS `Deal`. */
@Serializable
data class Deal(
    val id: String,
    @SerialName("product_id") val productId: String? = null,
    @SerialName("keyword_id") val keywordId: String? = null,
    @SerialName("matched_term") val matchedTerm: String? = null,
    val retailer: String,
    @SerialName("deal_title") val dealTitle: String,
    @SerialName("deal_price") val dealPrice: Double? = null,
    @SerialName("regular_price") val regularPrice: Double? = null,
    @SerialName("discount_pct") val discountPct: Double? = null,
    /** DATE column (`YYYY-MM-DD`), a calendar day in the company's timezone. */
    @SerialName("valid_from") val validFrom: String? = null,
    /** DATE column (`YYYY-MM-DD`), last valid calendar day. */
    @SerialName("valid_until") val validUntil: String? = null,
    @SerialName("image_url") val imageUrl: String? = null,
    @SerialName("image_url_large") val imageUrlLarge: String? = null,
    @SerialName("source_url") val sourceUrl: String? = null,
    @SerialName("external_url") val externalUrl: String? = null,
    @SerialName("matched_by") val matchedBy: String = "",
    val confidence: Double = 0.0,
    @SerialName("matched_tokens") val matchedTokens: List<String>? = null,
    @SerialName("requires_app") val requiresApp: Boolean = false,
    @SerialName("fetched_at") val fetchedAt: String? = null,
    @SerialName("offer_id") val offerId: String? = null,
    val products: DealProduct? = null,
    @SerialName("deal_keywords") val dealKeywords: DealKeywordMatch? = null,
)

/** Joined `products` relation on a deal row. */
@Serializable
data class DealProduct(
    val id: String? = null,
    val name: String? = null,
    @SerialName("image_path") val imagePath: String? = null,
    val sellprice: Double? = null,
)

/**
 * Keyword group embedded on keyword-matched rows. PostgREST returns the linked
 * products as `deal_keyword_products: [{ products: {...} }]`; [linkedProducts]
 * flattens that like iOS `DealKeywordMatch.linkedProducts`.
 */
@Serializable
data class DealKeywordMatch(
    val id: String? = null,
    val label: String? = null,
    val terms: List<String>? = null,
    @SerialName("deal_keyword_products") val dealKeywordProducts: List<LinkedProductRow> = emptyList(),
) {
    @Serializable
    data class LinkedProductRow(val products: DealProduct? = null)

    val linkedProducts: List<DealProduct>
        get() = dealKeywordProducts.mapNotNull { it.products }
}

/**
 * Per-user archive/pin annotation (`deal_user_state`), keyed by
 * (retailer, offer_id). Absence of a row means neither archived nor pinned.
 */
@Serializable
data class DealUserStateRow(
    val retailer: String,
    @SerialName("offer_id") val offerId: String,
    @SerialName("archived_at") val archivedAt: String? = null,
    @SerialName("pinned_at") val pinnedAt: String? = null,
)

/** Row of the `get_new_deal_keys` RPC. */
@Serializable
data class NewDealKeyRow(
    val retailer: String,
    @SerialName("offer_id") val offerId: String,
)

/** Per-product purchase-price aggregates from `get_product_purchase_summary`. Mirrors iOS `ProductPurchaseSummary`. */
@Serializable
data class ProductPurchaseSummary(
    @SerialName("product_id") val productId: String,
    @SerialName("ek_count") val ekCount: Int = 0,
    @SerialName("newest_net") val newestNet: Double? = null,
    @SerialName("newest_gross") val newestGross: Double? = null,
    @SerialName("newest_supplier") val newestSupplier: String? = null,
    @SerialName("newest_on") val newestOn: String? = null,
    @SerialName("min_gross") val minGross: Double? = null,
    @SerialName("min_supplier") val minSupplier: String? = null,
    @SerialName("min_on") val minOn: String? = null,
    @SerialName("max_gross") val maxGross: Double? = null,
    @SerialName("effective_tax_rate") val effectiveTaxRate: Double? = null,
)
